stratum/hal/lib/barefoot/p4runtime_bfrt_translator.cc

```diff
@@ -0,0 +1,742 @@
+    return request;
+  }
+  ::p4::v1::ReadRequest translated_request(request);
+  for (::p4::v1::Entity& entity : *translated_request.mutable_entities()) {
+    ASSIGN_OR_RETURN(entity, TranslateEntity(entity, /*to_sdk=*/true));
+  }
+  return translated_request;
+}
+
+::util::StatusOr<::p4::v1::ReadResponse>
+P4RuntimeBfrtTranslator::TranslateReadResponse(
+    const ::p4::v1::ReadResponse& response) {
+  absl::ReaderMutexLock l(&lock_);
+  if (!pipeline_require_translation_) {
+    return response;
+  }
+  ::p4::v1::ReadResponse translated_response(response);
+  for (::p4::v1::Entity& entity : *translated_response.mutable_entities()) {
+    ASSIGN_OR_RETURN(entity, TranslateEntity(entity, /*to_sdk=*/false));
+  }
+  return translated_response;
+}
+
+bool P4RuntimeBfrtTranslator::ReadResponseWriterWrapper::Write(
+    const ::p4::v1::ReadResponse& msg) {
+  auto status = p4runtime_bfrt_translator_->TranslateReadResponse(msg);
+  return status.ok() && writer_->Write(status.ConsumeValueOrDie());
+}
+
+bool P4RuntimeBfrtTranslator::StreamMessageResponseWriterWrapper::Write(
+    const ::p4::v1::StreamMessageResponse& msg) {
+  auto status = p4runtime_bfrt_translator_->TranslateStreamMessageResponse(msg);
+  return status.ok() && writer_->Write(status.ConsumeValueOrDie());
+}
+
+::util::Status P4RuntimeBfrtTranslator::PushChassisConfig(
+    const ChassisConfig& config, uint64 node_id) {
+  ::absl::WriterMutexLock l(&lock_);
+  // Port mapping for P4Runtime translation.
+  singleton_port_to_sdk_port_.clear();
+  sdk_port_to_singleton_port_.clear();
+  // Initialize with special ports.
+  singleton_port_to_sdk_port_[kSdnUnspecifiedPortId] = 0;
+  ASSIGN_OR_RETURN(const auto& cpu_sdk_port,
+                   bf_sde_interface_->GetPcieCpuPort(device_id_));
+  singleton_port_to_sdk_port_[kSdnCpuPortId] = cpu_sdk_port;
+  sdk_port_to_singleton_port_[cpu_sdk_port] = kSdnCpuPortId;
+  for (int pipe = 0; pipe < 4; pipe++) {
+    uint32 sdk_port = kTnaRecirculationPortBase | (pipe << 7);
+    uint32 sdn_port = kSdnTnaRecirculationPortBase + pipe;
+    singleton_port_to_sdk_port_[sdn_port] = sdk_port;
+    sdk_port_to_singleton_port_[sdk_port] = sdn_port;
+  }
+
+  for (const auto& singleton_port : config.singleton_ports()) {
+    uint32 singleton_port_id = singleton_port.id();
+    PortKey singleton_port_key(singleton_port.slot(), singleton_port.port(),
+                               singleton_port.channel());
+    ASSIGN_OR_RETURN(uint32 sdk_port_id,
+                     bf_sde_interface_->GetPortIdFromPortKey(
+                         device_id_, singleton_port_key));
+    singleton_port_to_sdk_port_[singleton_port_id] = sdk_port_id;
+    sdk_port_to_singleton_port_[sdk_port_id] = singleton_port_id;
+  }
+
+  return ::util::OkStatus();
+}
+
+::util::Status P4RuntimeBfrtTranslator::PushForwardingPipelineConfig(
+    const ::p4::config::v1::P4Info& p4info) {
+  ::absl::WriterMutexLock l(&lock_);
+  low_level_p4info_ = p4info;
```

Comment:
Generally, it's better to commit new state at the end after success. I.e. prepare a local low level map, then set it last.


See: https://github.com/stratum/stratum/blob/p4rt-translation/stratum/hal/lib/barefoot/bf_chassis_manager.cc#L582-L595